Refactor profiling

yinan1048576 opened the pull request:
Refactored core profiling codebase into two logical parts:

(a) `prof_data.c`: core internal data structure managing & dumping;
(b) `prof.c`: mutexes & outward-facing APIs.

Some internal functions had to be exposed out, but there are not
that many of them if the modularization is (hopefully) clean enough.

yinan1048576 commented:
Cleaned the `#include`s. Now the overlap of `#include` is minimal - only `assert.h` - which is a good sign.

davidtgoldblatt commented on Makefile.in:
```diff
@@ -117,6 +117,7 @@ C_SRCS := $(srcroot)src/jemalloc.c \
 	$(srcroot)src/pages.c \
 	$(srcroot)src/prng.c \
 	$(srcroot)src/prof.c \
+	$(srcroot)src/prof_data.c \
```
Ditto here as on the other diff - should be a corresponding MSVC project change.

yinan1048576 commented:
Added the new file to MSVC project. Also added comments to elaborate more about the separation of the two files.

BTW if there's new profiling functionality, unless it's a minor patch on an existing component, it'd be better added as a separate module (i.e. file) rather than appended to any of the two files. :)

yinan1048576 commented:
Rebase.

interwq commented:
Discussed offline. Reverted this to minimize the changes for 5.2.1. @yinan1048576 will need you to create another PR later
